### ISSUE TITLE
Connection callback does not  work as expected

### ISSUE THREAD

**Issue by archman** [pyepics/pyepics#163]
Hello,
It looks like the connection callback can only get connected event, the pyepics version I'm working with is `3.4.0rc1+26.g5585324`, it used to be working for `3.3.0`. Here is the code that can show this issue:

```python
from epics import PV

def conn_cb(pvname, conn, **kws): 
    print("{} is connected? {}".format(pvname, conn))

pv = PV('test_pv', connection_callback=conn_cb) 
```

And the simple IOC for testing (start with `softIoc -d test.db`):
```
# test.db
record(ao, "test_pv") {
    field(DTYP, "Soft Channel")
    field(OUT, "0")
    field(VAL, "0")
    field(PINI, "YES")
}
```
When the `pv` is created, it outputs `test_pv is connected? True`, but when the ioc is killed, no disconnection event was captured.

**Comment by archman** [comment]
And if killing the python session by `Ctrl + D` (after killing ioc session), will end up with `Segmentation fault` message.

**Comment by newville** [comment]
@archman @klauer  Thanks and yes, I can confirm that a connection callback does *not* get called on disconnect for 3.4.0rc2 either, but that it was called for 3.3.3.  FWIW, it does get called on *reconnect*.    And 3.4.0rc2 also appears to always segfault if there was a disconnection event.  

I think we should fix this before releasing 3.4.0.

**Comment by klauer** [comment]
Thanks, @archman! I'll look into it.

**Comment by dchabot** [comment]
Just looking at this now too.
Looks like something broke between 3.3.1 and 3.3.2 tags (ie tag 3.3.1 has the correct connection callback behavior).

**Comment by klauer** [comment]
It absolutely has to do with my cache rewrite, so I'm glad it was discovered before the release. 

I have some other reasonably significant pending changes (largely simplifications and atexit cleanup) in #161 but it's not quite there yet...

**Comment by dchabot** [comment]
@klauer, you're right, my mistake on the versioning where this was introduced.

**Comment by klauer** [comment]
No worries, @dchabot! I'm the one who should apologize here. Looks like I finally ran into the situation you were describing where the CA context was "bunk" (I think that was the word you had used to describe it, in any case ;) ) on the access rights callbacks.

Turns out that disconnection callbacks are also not called from a thread with a context, and that's the source of this problem. I pushed another fix/test based on that finding to #161. Further details are over there. Let's see how it works!
